### PR TITLE
Add managed Deflate64 support to ZipArchive

### PR DESCRIPTION
**Pull request by ianhays**
This PR adds back the old managed Deflate implementation and makes it compatible with the Deflate64 algorithm. ZipArchive is modified to use this managed Deflate64 stream whenever it detects an entry compressed using it. This doesn't add any compression support, only decompression, and an archive entry opened in Update mode will be written back using Deflate, not Deflate64.

This resolves the errors we've been seeing lately around windows zips >2gb throwing exceptions. resolves 
https://github.com/dotnet/corefx/issues/9925. Progress towards https://github.com/dotnet/corefx/issues/6602.

Tests coming in another PR where I'll also update the TestData version. I'll be waiting to merge this PR until that one is complete, but this one is big enough already.

@stephentoub

### CODE REVIEW

File: src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs [hunk 1]
```diff
@@ -25,153 +27,89 @@ public partial class DeflateStream : Stream
...
             _leaveOpen = leaveOpen;
-
-            _deflater = CreateDeflater(compressionLevel);
-
             _buffer = new byte[DefaultBufferSize];
         }
 
-        private static IDeflater CreateDeflater(CompressionLevel? compressionLevel)
+        /// <summary>
+        /// Sets up this DeflateManagedStream to be used for Zlib Deflation/Compression
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs]: This comment is counter-intuitive... the managed implementation is using zlib?

File: src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs [hunk 2]
```diff
@@ -43,6 +43,17 @@ internal DeflateManagedStream(Stream stream, bool leaveOpen, IFileFormatReader r
             InitializeInflater(stream, leaveOpen, reader);
         }
 
+        // A specific constructor to allow decompression of Deflate64
+        internal DeflateManagedStream(Stream stream, bool deflate64)
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs]: Seems potentially error-prone/confusing to use a bool, as the other ctors use a bool for something completely different, whether to close the stream or not when disposing.  Maybe a CompressionMethodValues?

File: src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 3]
```diff
@@ -425,6 +425,8 @@ private CompressionMethodValues CompressionMethod
             {
                 if (value == CompressionMethodValues.Deflate)
                     VersionToExtractAtLeast(ZipVersionNeededValues.Deflate);
+                if (value == CompressionMethodValues.Deflate64)
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: else if?

File: src/System.IO.Compression/src/System.IO.Compression.csproj [hunk 4]
```diff
@@ -34,6 +34,25 @@
     <Compile Include="$(CommonPath)\System\IO\PathInternal.cs">
       <Link>Common\System\IO\PathInternal.cs</Link>
     </Compile>
+    <Compile Include="System\IO\Compression\Deflate64\BlockType.cs" />
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: I assume this commit is basically just reverting the one where it was removed?

File: src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateStream.cs [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+                throw new InvalidOperationException(SR.CannotReadFromDeflateStream);
+        }
+
+        private void EnsureCompressionMode()
+        {
+            if (_mode != CompressionMode.Compress)
+                throw new InvalidOperationException(SR.CannotWriteToDeflateStream);
+        }
+
+        public override Task<int> ReadAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateStream.cs]: @joperezr had done some work to add Begin/End overrides to System.IO.Compression.  Does any of the code added back need those overrides added as well?

**Comment by stephentoub** [comment]
A few questions/comments, otherwise LGTM, assuming the new tests you added pass.

File: src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs (hunk 1 above)
Comment (ianhays) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs]: I built this DeflateStream off of the existing one so they'd be minimally different if we decided to merge them later. I remove this comment in a later commit.

File: src/System.IO.Compression/src/System.IO.Compression.csproj (hunk 4 above)
Comment (ianhays) [inline src/System.IO.Compression/src/System.IO.Compression.csproj]: Yes indeed! I didn't make any code changes.

File: src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateStream.cs (hunk 5 above)
Comment (ianhays) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateStream.cs]: If it's in the regular DeflateStream then it should probably be in the DeflateManagedStream as well. Are his changes [these in dev/api?](https://github.com/dotnet/corefx/blob/dev/api/src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs#L350) Is there anything blocking me from pulling those into this file and merging it into master?

@joperezr

**Pull request merged**